internal/schema/connection.go

```diff
@@ -0,0 +1,112 @@
+package schema
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func GetConnectionSchema(keyName string) *schema.Schema {
+	usernamePath := makePathRef(keyName, "username")
+	passwordPath := makePathRef(keyName, "password")
+	caFilePath := makePathRef(keyName, "ca_file")
+	caDataPath := makePathRef(keyName, "ca_data")
+	certFilePath := makePathRef(keyName, "cert_file")
+	certDataPath := makePathRef(keyName, "cert_data")
+	keyFilePath := makePathRef(keyName, "key_file")
+	keyDataPath := makePathRef(keyName, "key_data")
+
+	return &schema.Schema{
+		Description: "Elasticsearch connection configuration block.",
+		Type:        schema.TypeList,
+		MaxItems:    1,
+		Optional:    true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"username": {
+					Description: "Username to use for API authentication to Elasticsearch.",
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: schema.EnvDefaultFunc("ELASTICSEARCH_USERNAME", nil),
+					// RequiredWith: []string{passwordPath},
```

Comment:
Everything seems to carry over fine, apart from the `RequiredWith` validation for `username` and `password` fields.

Reading the docs I assumed this would only activate if either of those fields were present, but maybe the `DefaultFunc` is messing with it (`DefaultFunc` was only present on the elasticsearch schema, and not the `elasticsearch_connection` blocks)?

Reply:
https://github.com/hashicorp/terraform-plugin-sdk/issues/1099

Taking a look at the TF code there's not much we can do to have both of these attributes at the moment. Seems like this is the best approach for now, are you able to open an issue so if the TF issue is closed off we can improve the validation within the provider?

Reply:
Thanks for confirming mate, i'll create an issue and link it back

Reply:
I hooked this into the same logic we use for the env var, so we can still keep the validation for the connection blocks https://github.com/elastic/terraform-provider-elasticstack/pull/209/commits/e0ad01d84bac53fbc1b282be2735db20752fbc4c.

If you forsee any issues there, let me know and i'll back it out and create an issue instead.